### ISSUE TITLE
[YSQL] Prevent simultaneous CREATE INDEX/TABLE from creating multiple DocDB tables

### ISSUE THREAD

**Issue by jaki** [yugabyte/yugabyte-db#6269]
Running the same `CREATE INDEX` or `CREATE TABLE` in rapid succession (using different sessions) can result in duplicate docdb tables being created.  For easiest repro, see https://github.com/yugabyte/yugabyte-db/issues/6269#issuecomment-721474463 and https://github.com/yugabyte/yugabyte-db/issues/6269#issuecomment-721476241.

Here's an example with `CREATE INDEX` when backfill is enabled:

```sql
CREATE TABLE t (i int);
CREATE INDEX ON t (i);
```

At same time as above `CREATE INDEX`,

```sql
CREATE INDEX ON t (i);
```

Here's what happens, according to logs:

1. Create DocDB table for first index named `t_i_idx`
1. Create DocDB table for second index named `t_i_idx`
1. One of them gets error
1. The other goes on to backfill, completes, but doesn't get out of ALTERING state because the one that got error goes in `indexes_to_update`, so fully applied permissions aren't cleared, so `CatalogManager::GetTableSchema` gives `INDEX_PERM_WRITE_AND_DELETE`, so postgres `YBCPgWaitUntilIndexPermissionsAtLeast` keeps waiting until timeout

The error is

```
ERROR:  Catalog Version Mismatch: A DDL occurred while processing this query. Try again.    @     0x7f54dcb6ed2d  YBCGetStackTrace (src/yb/common/ybc_util.cc:287)
    @           0x9b229b  errmsg (src/postgres/src/backend/utils/error/elog.c:867)                                                                            
    @           0x866e84  YBPrepareCacheRefreshIfNeeded (src/postgres/src/backend/tcop/postgres.c:3844)                                                       
    @           0x86cead  PostgresMain (src/postgres/src/backend/tcop/postgres.c:4934)                                                                        
    @           0x7d92a2  BackendRun (src/postgres/src/backend/postmaster/postmaster.c:4428)                                                                  
    @           0x7d92a2  BackendStartup (src/postgres/src/backend/postmaster/postmaster.c:4094)                                                              
    @           0x7d92a2  ServerLoop (src/postgres/src/backend/postmaster/postmaster.c:1737)                                                                  
    @           0x7d92a2  PostmasterMain (src/postgres/src/backend/postmaster/postmaster.c:1400)                                                              
    @           0x72f644  PostgresServerProcessMain (src/postgres/src/backend/main/main.c:234)                                                                
    @           0x72f6e0                                                       
    @     0x7f54dbecd824  __libc_start_main (csu/libc-start.c:289)                                                                                            
    @           0x492b88  (unknown) (sysdeps/x86_64/start.S:118)                                                                                              
    @ 0xffffffffffffffff                                                                                                                                      
                                                                                                                                                              
DETAIL:  Internal error: Operation failed. Try again.: [Operation failed. Try again. (yb/docdb/conflict_resolution.cc:71): 039b6bc2-7d20-44d1-b813-f0f6298a0d5d Conflicts with higher priority transaction: bba7b69e-11ee-4ea4-a18d-d99c4a6ff7d3 (transaction error 3)]    @     0x7f54dcb6ed2d  YBCGetStackTrace (src/yb/com
mon/ybc_util.cc:287)                                                                                                                                          
    @           0x9b229b  errmsg (src/postgres/src/backend/utils/error/elog.c:867)                                                                            
    @           0x9d92eb  HandleYBStatusAtErrorLevel (src/postgres/src/backend/utils/misc/pg_yb_utils.c:325)                                                  
    @           0x9d931c  HandleYBStatus (src/postgres/src/backend/utils/misc/pg_yb_utils.c:309)                                                              
    @           0x9d9385  HandleYBStatusWithOwner (src/postgres/src/backend/utils/misc/pg_yb_utils.c:362)                                                     
    @           0x564ab3  ybcFetchNextHeapTuple (src/postgres/src/backend/access/ybc/ybcam.c:276)                                                             
    @           0x564ab3  ybc_getnext_heaptuple (src/postgres/src/backend/access/ybc/ybcam.c:1269)                                                            
    @           0x565064  ybc_systable_getnext (src/postgres/src/backend/access/ybc/ybcam.c:1374)                                                             
    @           0x4fe15f  systable_getnext (src/postgres/src/backend/access/index/genam.c:437)                                                                
    @           0x9a1fdc  RelationGetIndexList (src/postgres/src/backend/utils/cache/relcache.c:5026)                                                         
    @           0x6c8f3e  ExecOpenIndices (src/postgres/src/backend/executor/execIndexing.c:170)                                                              
    @           0x57a471  CatalogOpenIndexes (src/postgres/src/backend/catalog/indexing.c:53)                                                                 
    @           0x578fbe  AppendAttributeTuples (src/postgres/src/backend/catalog/index.c:570)                                                                
    @           0x578fbe  index_create (src/postgres/src/backend/catalog/index.c:1033)                                                                        
    @           0x6523a2  DefineIndex (src/postgres/src/backend/commands/indexcmds.c:1032)                                                                    
    @           0x873f62  ProcessUtilitySlow (src/postgres/src/backend/tcop/utility.c:1435)                                                                   
    @           0x87306c  standard_ProcessUtility (src/postgres/src/backend/tcop/utility.c:962)                                                               
    @           0x8730fa  YBProcessUtilityDefaultHook (src/postgres/src/backend/tcop/utility.c:3574)                                                          
    @     0x7f54cfa23a50  pgss_ProcessUtility (src/postgres/contrib/pg_stat_statements/pg_stat_statements.c:1056)                                             
    @     0x7f54cf81c091  ybpgm_ProcessUtility (src/postgres/contrib/yb_pg_metrics/yb_pg_metrics.c:534)                                                       
    @     0x7f54cf614a12  pgaudit_NextProcessUtility_hook (src/postgres/contrib/pgaudit/pgaudit.c:1225)                                                       
    @     0x7f54cf6160aa  pgaudit_ProcessUtility_hook (src/postgres/contrib/pgaudit/pgaudit.c:1253)                                                           
    @           0x9dacf5  YBTxnDdlProcessUtility (src/postgres/src/backend/utils/misc/pg_yb_utils.c:1148)                                                     
    @           0x8731c8  ProcessUtility (src/postgres/src/backend/tcop/utility.c:373)                                                                        
    @           0x86f4fc  PortalRunUtility (src/postgres/src/backend/tcop/pquery.c:1189)                                                                      
    @           0x86fa94  PortalRunMulti (src/postgres/src/backend/tcop/pquery.c:1349)                                                                        
    @           0x870a74  PortalRun (src/postgres/src/backend/tcop/pquery.c:810)                                                                              
    @           0x86a58d  exec_simple_query (src/postgres/src/backend/tcop/postgres.c:1161)                                                                   
    @           0x86a58d  yb_exec_simple_query (src/postgres/src/backend/tcop/postgres.c:4313)                                                                
    @           0x86ce4a  PostgresMain (src/postgres/src/backend/tcop/postgres.c:4924)                                                                        
    @           0x7d92a2  BackendRun (src/postgres/src/backend/postmaster/postmaster.c:4428)                                                                  
    @           0x7d92a2  BackendStartup (src/postgres/src/backend/postmaster/postmaster.c:4094)                                                              
    @           0x7d92a2  ServerLoop (src/postgres/src/backend/postmaster/postmaster.c:1737)                                                                  
    @           0x7d92a2  PostmasterMain (src/postgres/src/backend/postmaster/postmaster.c:1400)                                                              
    @           0x72f644  PostgresServerProcessMain (src/postgres/src/backend/main/main.c:234)                                                                
    @           0x72f6e0                                                                                                                                      
    @     0x7f54dbecd824  __libc_start_main (csu/libc-start.c:289)                                                                                            
    @           0x492b88  (unknown) (sysdeps/x86_64/start.S:118)                                                                                              
    @ 0xffffffffffffffff                                                                                                                                      
    @     0x7f54dcb6ed2d  YBCGetStackTrace (src/yb/common/ybc_util.cc:287)                                                                                    
    @           0x9b27c7  errdetail (src/postgres/src/backend/utils/error/elog.c:948)                                                                         
    @           0x866e70  YBPrepareCacheRefreshIfNeeded (src/postgres/src/backend/tcop/postgres.c:3844)                                                       
    @           0x86cead  PostgresMain (src/postgres/src/backend/tcop/postgres.c:4934)                                                                        
    @           0x7d92a2  BackendRun (src/postgres/src/backend/postmaster/postmaster.c:4428)                                                                  
    @           0x7d92a2  BackendStartup (src/postgres/src/backend/postmaster/postmaster.c:4094)                                                              
    @           0x7d92a2  ServerLoop (src/postgres/src/backend/postmaster/postmaster.c:1737)                                                                  
    @           0x7d92a2  PostmasterMain (src/postgres/src/backend/postmaster/postmaster.c:1400)                                                              
    @           0x72f644  PostgresServerProcessMain (src/postgres/src/backend/main/main.c:234)                                                                
    @           0x72f6e0                                                       
    @     0x7f54dbecd824  __libc_start_main (csu/libc-start.c:289)             
    @           0x492b88  (unknown) (sysdeps/x86_64/start.S:118)               
    @ 0xffffffffffffffff
```

Now, there are two docdb index tables and one invalid postgres index table.  The orphaned docdb index will have to be manually deleted (e.g. by `yb-admin`).

**Comment by jaki** [comment]
`CREATE TABLE` doesn't have this issue because one of the requests fails earlier before creating DocDB table:

```
ERROR:  Operation failed. Try again.: [Operation failed. Try again. (yb/docdb/conflict_resolution.cc:71): 13e87704-1667-4c25-a9e0-5880a9fbb14c Conflicts with higher priority transaction: 83051f52-7c7f-4b73-996e-028f583593cf (transaction error 3)]    @     0x7fd3e89b2d2d  YBCGetStackTrace (src/yb/common/ybc_util.cc:2
87)                                                                                                                                                                                                                                                                                                                          
    @           0x9b229b  errmsg (src/postgres/src/backend/utils/error/elog.c:867)                                                                                                                                                                                                                                           
    @           0x9d92eb  HandleYBStatusAtErrorLevel (src/postgres/src/backend/utils/misc/pg_yb_utils.c:325)                                                                                                                                                                                                                 
    @           0x9d931c  HandleYBStatus (src/postgres/src/backend/utils/misc/pg_yb_utils.c:309)                                                                                                                                                                                                                             
    @           0x9d9385  HandleYBStatusWithOwner (src/postgres/src/backend/utils/misc/pg_yb_utils.c:362)                                                                                                                                                                                                                    
    @           0x564a73  ybcFetchNextHeapTuple (src/postgres/src/backend/access/ybc/ybcam.c:269)                                                                                                                                                                                                                            
    @           0x564a73  ybc_getnext_heaptuple (src/postgres/src/backend/access/ybc/ybcam.c:1269)                                                                                                                                                                                                                           
    @           0x565064  ybc_systable_getnext (src/postgres/src/backend/access/ybc/ybcam.c:1374)                                                                                                                                                                                                                            
    @           0x4fe15f  systable_getnext (src/postgres/src/backend/access/index/genam.c:437)                                                                                                                                                                                                                               
    @           0x98e844  SearchCatCacheMiss (src/postgres/src/backend/utils/cache/catcache.c:1760)                                                                                                                                                                                                                          
    @           0x98fc9a  SearchCatCacheInternal (src/postgres/src/backend/utils/cache/catcache.c:1655)                                                                                                                                                                                                                      
    @           0x98fc9a  SearchCatCache (src/postgres/src/backend/utils/cache/catcache.c:1509)                                                               
    @           0x9a7e55  SearchSysCache (src/postgres/src/backend/utils/cache/syscache.c:1582)                                                               
    @           0x9a803b  SearchSysCacheExists (src/postgres/src/backend/utils/cache/syscache.c:1680)                                                         
    @           0x5a7913  makeArrayTypeName (src/postgres/src/backend/catalog/pg_type.c:792)                                                                  
    @           0x570f09  heap_create_with_catalog (src/postgres/src/backend/catalog/heap.c:1259)                                                             
    @           0x686b22  DefineRelation (src/postgres/src/backend/commands/tablecmds.c:832)                                                                  
    @           0x87338a  ProcessUtilitySlow (src/postgres/src/backend/tcop/utility.c:1042)                                                                   
    @           0x87306c  standard_ProcessUtility (src/postgres/src/backend/tcop/utility.c:962)                                                               
    @           0x8730fa  YBProcessUtilityDefaultHook (src/postgres/src/backend/tcop/utility.c:3574)                                                          
    @     0x7fd3db867a50  pgss_ProcessUtility (src/postgres/contrib/pg_stat_statements/pg_stat_statements.c:1056)                                             
    @     0x7fd3db660091  ybpgm_ProcessUtility (src/postgres/contrib/yb_pg_metrics/yb_pg_metrics.c:534)                                                       
    @     0x7fd3db458a12  pgaudit_NextProcessUtility_hook (src/postgres/contrib/pgaudit/pgaudit.c:1225)                                                       
    @     0x7fd3db45a0aa  pgaudit_ProcessUtility_hook (src/postgres/contrib/pgaudit/pgaudit.c:1253)                                                           
    @           0x9dacf5  YBTxnDdlProcessUtility (src/postgres/src/backend/utils/misc/pg_yb_utils.c:1148)                                                     
    @           0x8731c8  ProcessUtility (src/postgres/src/backend/tcop/utility.c:373)                                                                        
    @           0x86f4fc  PortalRunUtility (src/postgres/src/backend/tcop/pquery.c:1189)                                                                      
    @           0x86fa94  PortalRunMulti (src/postgres/src/backend/tcop/pquery.c:1349)                                                                        
    @           0x870a74  PortalRun (src/postgres/src/backend/tcop/pquery.c:810)                                                                              
    @           0x86a58d  exec_simple_query (src/postgres/src/backend/tcop/postgres.c:1161)                                                                   
    @           0x86a58d  yb_exec_simple_query (src/postgres/src/backend/tcop/postgres.c:4313)                                                                
    @           0x86ce4a  PostgresMain (src/postgres/src/backend/tcop/postgres.c:4924)                                                                        
    @           0x7d92a2  BackendRun (src/postgres/src/backend/postmaster/postmaster.c:4428)                                                                  
    @           0x7d92a2  BackendStartup (src/postgres/src/backend/postmaster/postmaster.c:4094)                                                              
    @           0x7d92a2  ServerLoop (src/postgres/src/backend/postmaster/postmaster.c:1737)                                                                  
    @           0x7d92a2  PostmasterMain (src/postgres/src/backend/postmaster/postmaster.c:1400)                                                              
    @           0x72f644  PostgresServerProcessMain (src/postgres/src/backend/main/main.c:234)                                                                
    @           0x72f6e0                                                       
    @     0x7fd3e7d11824  __libc_start_main (csu/libc-start.c:289)             
    @           0x492b88  (unknown) (sysdeps/x86_64/start.S:118)               
    @ 0xffffffffffffffff
```

`heap_create_with_catalog` is called and errors before `YBCCreateTable`.

**Comment by jaki** [comment]
`CREATE INDEX` without backfill has the same problem.  The timing is more strict, so try

```sh
for i in {1..10}; do
  ./bin/ysqlsh -c "CREATE INDEX ON t (i)" &;
done
```

**Comment by jaki** [comment]
Actually, it does seem possible for `CREATE TABLE`.  Trying

```sh
for i in {1..10}; do
  ./bin/ysqlsh -c "CREATE TABLE t (i int)" &;
done
```

I tend to get 2 docdb tables.  Trying `{1..50}`, I got 3 docdb tables.

**Comment by jaki** [comment]
This looks resolved by Nicolas's commit 9f024d211d245aafa2c4bd24954b2e76f7afcc7e.
